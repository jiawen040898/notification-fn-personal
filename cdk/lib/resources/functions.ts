import type { Function as AwsLambdaFunction } from 'aws-cdk-lib/aws-lambda';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';

import { BaseFunction } from '../base';
import type {
    NotificationProcessEventMessageNotificationFnConfig,
    NotificationPushNotificationFnConfig,
    NotificationSendEmailFnConfig,
} from '../interfaces';
import { accountId, version } from '../variables';
import type { IAMRoleGroupResources } from './iam/iam-roles';
import type { LayerGroupResources } from './layers';
import type SQSGroupResources from './sqs';

/**
 * FunctionGroupResourcesProps
 *
 * @param iamRoleGroupResources {@link IAMRoleGroupResources}
 * @param sqsGroupResources {@link SQSGroupResources}
 * @param layerGroupResources {@link LayerGroupResources}
 */
type FunctionGroupResourcesProps = {
    iamRoleGroupResources: IAMRoleGroupResources;
    sqsGroupResources: SQSGroupResources;
    layerGroupResources: LayerGroupResources;
};

export class FunctionGroupResources extends Construct {
    public readonly notificationHandleBounceNotificationFn: AwsLambdaFunction;
    public readonly notificationHandleNotificationMetricFn: AwsLambdaFunction;
    public readonly notificationProcessEventMessageNotificationFn: AwsLambdaFunction;
    public readonly notificationSendEmailFn: AwsLambdaFunction;
    public readonly notificationPushNotificationFn: AwsLambdaFunction;

    /**
     * FunctionGroupResources
     *
     * @public notificationHandleBounceNotificationFn {@link AwsLambdaFunction}
     * @public notificationHandleNotificationMetricFn {@link AwsLambdaFunction}
     * @public notificationProcessEventMessageNotificationFn {@link AwsLambdaFunction}
     * @public notificationSendEmailFn {@link AwsLambdaFunction}
     * @public notificationPushNotificationFn {@link AwsLambdaFunction}
     *
     * @param scope {@link Construct}
     * @param id
     * @param props {@link FunctionGroupResourcesProps}
     */
    constructor(
        scope: Construct,
        id: string,
        props: FunctionGroupResourcesProps,
    ) {
        super(scope, id);

        this.notificationHandleBounceNotificationFn = new BaseFunction(
            this,
            'notification-handleBounceNotification-fn',
            {
                functionName: 'notification-handleBounceNotification-fn',
                description: `Update Email Activity Message Bounce Status by process SES Notification from SNS (v${version})`,
                entry: 'src/functions/handle-bounce-notification.ts',
                // reservedConcurrentExecutions: 25,
                isLogGroupExists: true,
                iamRole: props.iamRoleGroupResources.notificationLambdaRole,
                layers: [props.layerGroupResources.notificationFnLayer],
                snsEventSources: [
                    {
                        topic: Topic.fromTopicArn(
                            this,
                            'ses-email-bounce',
                            `arn:aws:sns:us-west-2:${accountId}:ses-email-bounce`,
                        ),
                    },
                ],
            },
        ).lambda;

        this.notificationHandleNotificationMetricFn = new BaseFunction(
            this,
            'notification-handleNotificationMetric-fn',
            {
                functionName: 'notification-handleNotificationMetric-fn',
                description: `Update Email Activity Message Metrics (Open/Click) by process SES Notification from SNS (v${version})`,
                entry: 'src/functions/handle-notification-metric.ts',
                // reservedConcurrentExecutions: 25,
                isLogGroupExists: true,
                iamRole: props.iamRoleGroupResources.notificationLambdaRole,
                layers: [props.layerGroupResources.notificationFnLayer],
            },
        ).lambda;

        this.notificationProcessEventMessageNotificationFn =
            new BaseFunction<NotificationProcessEventMessageNotificationFnConfig>(
                this,
                'notification-processEventMessageNotification-fn',
                {
                    functionName:
                        'notification-processEventMessageNotification-fn',
                    description: `Process Event Notification (v${version})`,
                    entry: 'src/functions/process-event-message-notification.ts',
                    // reservedConcurrentExecutions: 25,
                    isLogGroupExists: true,
                    lambdaSpecificEnvironmentVariables: {
                        NOTIFICATION_EMAIL_REQUEST_QUEUE_URL: `${StringParameter.valueForStringParameter(
                            scope,
                            '/configs/api/AWS_SQS_BASE_DNS',
                        )}notification-email-request-queue`,
                        PUSH_NOTIFICATION_REQUEST_QUEUE_URL: `${StringParameter.valueForStringParameter(
                            scope,
                            '/configs/api/AWS_SQS_BASE_DNS',
                        )}notification-push-notification-queue`,
                    },
                    iamRole: props.iamRoleGroupResources.notificationLambdaRole,
                    layers: [props.layerGroupResources.notificationFnLayer],
                    sqsEventSources: [
                        {
                            queue: props.sqsGroupResources
                                .notificationDomainQueue,
                            sqsEventSourceProps: { batchSize: 1 },
                        },
                    ],
                },
            ).lambda;

        this.notificationSendEmailFn =
            new BaseFunction<NotificationSendEmailFnConfig>(
                this,
                'notification-sendEmail-fn',
                {
                    functionName: 'notification-sendEmail-fn',
                    description: `Send Email via SES by process Email Request from SQS (v${version})`,
                    entry: 'src/functions/send-email.ts',
                    // reservedConcurrentExecutions: 25,
                    isLogGroupExists: true,
                    lambdaSpecificEnvironmentVariables: {
                        SES_REGION: 'us-west-2',
                    },
                    iamRole: props.iamRoleGroupResources.notificationLambdaRole,
                    layers: [props.layerGroupResources.notificationFnLayer],
                    sqsEventSources: [
                        {
                            queue: props.sqsGroupResources
                                .notificationEmailRequestQueue,
                            sqsEventSourceProps: { batchSize: 1 },
                        },
                    ],
                },
            ).lambda;

        this.notificationPushNotificationFn =
            new BaseFunction<NotificationPushNotificationFnConfig>(
                this,
                'notification-push-notification-fn',
                {
                    functionName: 'notification-push-notification-fn',
                    description: `Send push notification via Pusher from SQS (v${version})`,
                    entry: 'src/functions/send-push-notification.ts',
                    // reservedConcurrentExecutions: 25,
                    isLogGroupExists: true,
                    lambdaSpecificEnvironmentVariables: {
                        PUSHER_SM_NAME: 'pusher-credentials',
                    },
                    iamRole: props.iamRoleGroupResources.notificationLambdaRole,
                    layers: [props.layerGroupResources.notificationFnLayer],
                    sqsEventSources: [
                        {
                            queue: props.sqsGroupResources
                                .notificationPushNotificationQueue,
                            sqsEventSourceProps: { batchSize: 1 },
                        },
                    ],
                },
            ).lambda;
    }
}
