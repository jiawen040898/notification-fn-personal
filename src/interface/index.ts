export * from './company.interface';
export * from './email-request';
export * from './event-model.interface';
export * from './feedback.interface';
export * from './identity.interface';
export * from './manager-subordinate-aggregated-report-data.interface';
export * from './program.interface';
export * from './push-notification-request.interface';
export * from './ses-notification';
