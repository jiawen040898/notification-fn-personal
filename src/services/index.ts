export * from './app-notification.service';
export * from './company.service';
export * from './company-email-template.service';
export * from './email-notification.service';
export * from './email-tracking.service';
export * from './identity.service';
export * from './push-notification.service';
export * from './send-email.service';
export * from './talent-management-notification.service';
