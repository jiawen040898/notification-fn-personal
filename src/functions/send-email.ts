import { ErrorDetails, sqsRecordUtil } from '@pulsifi/fn';
import { Handler, SQSEvent } from 'aws-lambda';

import { getDataSource } from '../database';
import { IEmailRequest, ITemplateVariable } from '../interface';
import { eventMiddleware } from '../middleware';
import { SendEmailService } from '../services';
import { validateUrl } from '../utils';

const handleEvent: Handler = async (event: SQSEvent) => {
    const dataSource = await getDataSource();
    const sendEmailService = new SendEmailService(dataSource);
    for (const record of event.Records) {
        const emailRequest = sqsRecordUtil.parseBody<IEmailRequest>(record);
        validateEmailTemplateVariableLink(emailRequest.variables);

        await sendEmailService.sendEmail(emailRequest);
    }
};

/** check if any email template variable key end with 'link' or 'url',
 *  if yes and content exist, make sure they are valid url,
 *  prompt error if they are not valid url
 */
export function validateEmailTemplateVariableLink(
    variables: Partial<ITemplateVariable>,
): void {
    for (const key in variables) {
        if (key.endsWith('link') || key.endsWith('url')) {
            const url = variables[key as keyof ITemplateVariable] as string;

            if (url && !validateUrl(url)) {
                throw new EmailVariableValidationException({
                    error_codes: ['invalid_url_format'],
                    key,
                    url,
                });
            }
        }
    }
}

class EmailVariableValidationException extends Error {
    errorDetails: ErrorDetails;

    constructor(errorDetails: ErrorDetails) {
        super(`Invalid URL format for variable key: ${errorDetails.key}`);
        this.errorDetails = errorDetails;
    }
}

export const handler = eventMiddleware(handleEvent);
