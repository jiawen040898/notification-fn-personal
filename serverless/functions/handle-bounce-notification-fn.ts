import { AwsFunctionHandler } from 'serverless/aws';

import { layers, tags, version } from './variables';

export const handleBounceNotification: AwsFunctionHandler = {
    name: 'notification-handleBounceNotification-fn',
    description: `Update Email Activity Message Bounce Status by process SES Notification from SNS (v${version})`,
    handler: 'src/functions/handle-bounce-notification.handler',
    // reservedConcurrency: 25,
    events: [
        {
            sns: {
                arn: 'arn:aws:sns:us-west-2:${aws:accountId}:ses-email-bounce',
            },
        },
    ],
    layers,
    tags,
};
