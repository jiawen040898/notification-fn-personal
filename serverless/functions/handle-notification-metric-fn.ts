import { AwsFunctionHandler } from 'serverless/aws';

import { layers, tags, version } from './variables';

export const handleNotificationMetric: AwsFunctionHandler = {
    name: 'notification-handleNotificationMetric-fn',
    description: `Update Email Activity Message Metrics (Open/Click) by process SES Notification from SNS (v${version})`,
    handler: 'src/functions/handle-notification-metric.handler',
    // reservedConcurrency: 25,
    // events: [
    //     {
    //         sns: {
    //             arn: 'arn:aws:sns:us-west-2:${aws:accountId}:ses-tracking-topic',
    //         },
    //     },
    // ],
    layers,
    tags,
};
