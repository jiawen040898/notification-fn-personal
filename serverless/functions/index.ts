/* eslint-disable @typescript-eslint/naming-convention */
import { Functions } from 'serverless/aws';

import { handleBounceNotification } from './handle-bounce-notification-fn';
import { handleNotificationMetric } from './handle-notification-metric-fn';
import { processEventMessageNotification } from './process-event-message-notification-fn';
import { sendEmail } from './send-email-fn';
import { sendPushNotification } from './send-push-notification-fn';

export const functions: Functions = {
    sendEmail,
    sendPushNotification,
    processEventMessageNotification,
    handleBounceNotification,
    handleNotificationMetric,
};
