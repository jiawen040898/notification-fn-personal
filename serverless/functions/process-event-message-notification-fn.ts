import { AwsFunctionHandler } from 'serverless/aws';

import { layers, tags, version } from './variables';

export const processEventMessageNotification: AwsFunctionHandler = {
    name: 'notification-processEventMessageNotification-fn',
    description: `Process Event Notification (v${version})`,
    handler: 'src/functions/process-event-message-notification.handler',
    // reservedConcurrency: 25,
    environment: {
        NOTIFICATION_EMAIL_REQUEST_QUEUE_URL:
            '${ssm:/configs/api/AWS_SQS_BASE_DNS}notification-email-request-queue',
        PUSH_NOTIFICATION_REQUEST_QUEUE_URL:
            '${ssm:/configs/api/AWS_SQS_BASE_DNS}notification-push-notification-queue',
    },
    events: [
        {
            sqs: {
                arn: 'arn:aws:sqs:${aws:region}:${aws:accountId}:notification-domain-queue.fifo',
                batchSize: 1,
            },
        },
    ],
    layers,
    tags,
};
