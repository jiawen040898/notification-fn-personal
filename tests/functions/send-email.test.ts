import { dateTimeUtil, generatorUtil, objectParser } from '@pulsifi/fn';
import { DataSource } from 'typeorm';

import {
    DefaultLocale,
    EmailTemplateCommunicationType,
    EmailTemplateUsage,
    Locale,
} from '../../src/constants';
import { validateEmailTemplateVariableLink } from '../../src/functions/send-email';
import { IEmailRequest } from '../../src/interface';
import {
    EmailActivityMessage,
    EmailActivityMessageDeliveryStatus,
    EmailActivityMessageRecipientGroup,
    EmailTemplate,
} from '../../src/models';
import { SendEmailService } from '../../src/services';
import { testData } from '../fixtures';
import { getTestDataSourceAndAddData } from '../setup/test-database.setup';

const mockSESClient = {
    send: jest.fn(),
};

jest.mock('../../src/services/company.service', () => {
    const mBInstance = {
        getCompanyData: jest.fn(() => [testData.companyDataResponse]),
        getCompanyLocaleByCompanyId: jest.fn(() => [DefaultLocale]),
    };
    const mB = jest.fn(() => mBInstance);
    return { CompanyService: mB };
});

jest.mock('@aws-sdk/client-ses', () => {
    return {
        ...jest.requireActual('@aws-sdk/client-ses'),
        SESClient: jest.fn(() => mockSESClient),
    };
});

jest.mock('@pulsifi/fn/utils/logger.util');

describe('sendEmail', () => {
    let dataSource: DataSource;
    const sesMessageId = generatorUtil.uuid();

    beforeAll(async () => {
        const emailTemplate: EmailTemplate = {
            id: 1,
            company_id: 1,
            subject: 'Hello world {{job_title}}',
            sender_name: '{{sender_name}} via Pulsifi',
            email_communication_type:
                EmailTemplateCommunicationType.APPLICATION_SHORTLIST,
            body_in_html: '<html>Hello {{first_name}} {{year}}</html>',
            usage: EmailTemplateUsage.CLIENT,
            language: Locale.EN_US,
            is_deleted: false,
            created_by: 1,
            updated_by: 1,
        };

        const systemFeedbackEmailTemplate: EmailTemplate = {
            id: 2,
            company_id: 0,
            subject: 'Hello world {{feedback_cycle.name}}',
            sender_name: '{{sender_name}} via Pulsifi',
            email_communication_type:
                EmailTemplateCommunicationType.EMPLOYEE_FEEDBACK_NOMINATION_REQUESTED,
            body_in_html: '<html>Hello {{first_name}} {{year}}</html>',
            usage: EmailTemplateUsage.SYSTEM,
            language: Locale.EN_US,
            is_deleted: false,
            created_by: 1,
            updated_by: 1,
        };

        const emailLog: EmailActivityMessage = {
            id: 1,
            company_id: 1,
            email_template_id: 1,
            delivery_status: EmailActivityMessageDeliveryStatus.SENT,
            subject: 'Hello world MT Program',
            sender_email: 'sender@domain.com',
            sender_name: 'Sender',
            email_communication_type: 'application_shortlisted',
            body_in_html: `<html>Hello David ${dateTimeUtil.getCurrentYear()}</html>`,
            recipient_email: 'da@da.com',
            recipient_group: 'user',
            body_variables: objectParser.toJSON({}),
            recipient_id: sesMessageId,
            ext_ses_message_id: sesMessageId,
            open_count: 0,
            click_count: 0,
            is_deleted: false,
            sent_at: new Date(),
            created_by: 1,
            updated_by: 1,
        };

        dataSource = await getTestDataSourceAndAddData([
            {
                entityClass: EmailTemplate,
                data: [emailTemplate, systemFeedbackEmailTemplate],
            },
            {
                entityClass: EmailActivityMessage,
                data: emailLog,
            },
        ]);
    });

    afterAll(() => {
        dataSource.destroy();
        jest.resetAllMocks();
    });

    beforeEach(() => {
        mockSESClient.send.mockReset();
    });

    it('should pass sending email', async () => {
        // Arrange
        mockSESClient.send.mockResolvedValue({
            data: {
                MessageId: 1,
            },
        });

        const data: IEmailRequest = {
            recipient_id: sesMessageId,
            recipient_email: 'dav@emai.com',
            recipient_group: EmailActivityMessageRecipientGroup.USER,
            email_template_id: 1,
            variables: {
                first_name: 'David',
                job_title: 'MT Program',
                sender_name: 'Nestl√©',
            },
        };

        // Act
        await new SendEmailService(dataSource).sendEmail(data);

        // Assert
        const foundEmailLog = await dataSource
            .getRepository(EmailActivityMessage)
            .findOneBy({ recipient_id: data.recipient_id });

        expect(foundEmailLog?.subject).toEqual('Hello world MT Program');
        expect(foundEmailLog?.body_in_html).toEqual(
            `<html>Hello David ${dateTimeUtil.getCurrentYear()}</html>`,
        );
        expect(foundEmailLog?.delivery_status).toEqual(
            EmailActivityMessageDeliveryStatus.SENT,
        );
        expect(foundEmailLog?.sent_at).not.toBeNull();
        expect(mockSESClient.send.mock.calls).toMatchSnapshot();
    });

    it('should fail sending email with ses error', async () => {
        // Arrange
        mockSESClient.send.mockRejectedValue(new Error('SES Fail'));

        const data: IEmailRequest = {
            recipient_id: generatorUtil.uuid(),
            recipient_email: 'dav@emai.com',
            recipient_group: EmailActivityMessageRecipientGroup.USER,
            email_template_id: 1,
            variables: {
                first_name: 'David',
            },
        };

        // Act
        const action = async () => {
            await new SendEmailService(dataSource).sendEmail(data);
        };

        // Assert
        await expect(action()).rejects.toThrow(Error);
    });

    it('should pass sending email with email template data ', async () => {
        // Arrange
        mockSESClient.send.mockResolvedValue({
            data: {
                MessageId: 1,
            },
        });

        const data: IEmailRequest = {
            company_id: 9,
            recipient_id: sesMessageId,
            recipient_email: 'dav@emai.com',
            recipient_group: EmailActivityMessageRecipientGroup.USER,
            // email_template_id: 1,
            email_communication_type:
                EmailTemplateCommunicationType.EMPLOYEE_FEEDBACK_NOMINATION_REQUESTED,
            variables: {
                cycle: {
                    name: 'Q1 FC',
                },
                receiver: {
                    first_name: 'Benny',
                    last_name: 'Ng',
                    full_name: 'Benny Ng',
                    user_account_id: 1,
                    employee_id: 'Anything',
                    email: 'da@da.com',
                    id: 'Anything',
                },
            },
            email_template: {
                subject: 'Please submit your nomination for {{cycle.name}}',
                body_in_html: '<html>Hello {{reviewe.name}}</html>',
            },
        };

        // Act
        await new SendEmailService(dataSource).sendEmail(data);

        // Assert
        const foundEmailLog = await dataSource
            .getRepository(EmailActivityMessage)
            .findOneBy({ recipient_id: data.recipient_id });

        expect(foundEmailLog?.subject).toMatchSnapshot();
        expect(foundEmailLog?.body_in_html).toMatchSnapshot();
        expect(foundEmailLog?.delivery_status).toEqual(
            EmailActivityMessageDeliveryStatus.SENT,
        );
        expect(foundEmailLog?.sent_at).not.toBeNull();
        expect(mockSESClient.send.mock.calls).toMatchSnapshot();
    });

    describe('validateEmailTemplateVariableLink', () => {
        it('should throw error if url is invalid', () => {
            // Arrange
            const variables = {
                employee_app_url: 'invalid_url',
            };

            // Act
            const action = () => {
                validateEmailTemplateVariableLink(variables);
            };

            // Assert
            expect(action).toThrowErrorMatchingSnapshot();
        });

        it('should throw error if link is invalid', () => {
            // Arrange
            const variables = {
                invite_link: 'invalid_url',
            };

            // Act
            const action = () => {
                validateEmailTemplateVariableLink(variables);
            };

            // Assert
            expect(action).toThrowErrorMatchingSnapshot();
        });

        it('should not return any error if url and links are fine', () => {
            // Arrange
            const variables = {
                employee_app_url: 'https://www.google.com',
                invite_link: 'https://www.google.com',
                company_id: 5,
            };

            // Act
            const action = () => {
                validateEmailTemplateVariableLink(variables);
            };

            // Assert
            expect(action).not.toThrow();
        });
    });
});
