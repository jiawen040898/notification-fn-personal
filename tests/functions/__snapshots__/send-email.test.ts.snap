// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendEmail should pass sending email 1`] = `
[
  [
    SendEmailCommand {
      "input": {
        "Destination": {
          "ToAddresses": [
            "dav@emai.com",
          ],
        },
        "Message": {
          "Body": {
            "Html": {
              "Charset": "UTF-8",
              "Data": "<html>Hello David 2024</html>",
            },
          },
          "Subject": {
            "Charset": "UTF-8",
            "Data": "Hello world MT Program",
          },
        },
        "ReplyToAddresses": [],
        "Source": "=?utf-8?Q?Nestl=C3=A9?= via Pulsifi <no-reply@pulsifi.me>",
        "Tags": [
          {
            "Name": "emailCommunicationType",
            "Value": "application_shortlist",
          },
        ],
      },
      "middlewareStack": {
        "add": [Function],
        "addRelativeTo": [Function],
        "applyToStack": [Function],
        "clone": [Function],
        "concat": [Function],
        "identify": [Function],
        "remove": [Function],
        "removeByTag": [Function],
        "resolve": [Function],
        "use": [Function],
      },
    },
  ],
]
`;

exports[`sendEmail should pass sending email with email template data  1`] = `"Hello world MT Program"`;

exports[`sendEmail should pass sending email with email template data  2`] = `"<html>Hello David 2024</html>"`;

exports[`sendEmail should pass sending email with email template data  3`] = `
[
  [
    SendEmailCommand {
      "input": {
        "ConfigurationSetName": "tracking",
        "Destination": {
          "ToAddresses": [
            "dav@emai.com",
          ],
        },
        "Message": {
          "Body": {
            "Html": {
              "Charset": "UTF-8",
              "Data": "<html>Hello </html>",
            },
          },
          "Subject": {
            "Charset": "UTF-8",
            "Data": "Please submit your nomination for Q1 FC",
          },
        },
        "ReplyToAddresses": [],
        "Source": " via Pulsifi <no-reply@pulsifi.me>",
        "Tags": [
          {
            "Name": "emailCommunicationType",
            "Value": "employee_feedback_nomination_requested",
          },
        ],
      },
      "middlewareStack": {
        "add": [Function],
        "addRelativeTo": [Function],
        "applyToStack": [Function],
        "clone": [Function],
        "concat": [Function],
        "identify": [Function],
        "remove": [Function],
        "removeByTag": [Function],
        "resolve": [Function],
        "use": [Function],
      },
    },
  ],
]
`;

exports[`sendEmail validateEmailTemplateVariableLink should throw error if link is invalid 1`] = `"Invalid URL format for variable key: invite_link"`;

exports[`sendEmail validateEmailTemplateVariableLink should throw error if url is invalid 1`] = `"Invalid URL format for variable key: employee_app_url"`;
